src/async_fuse/memfs/cache/memory_cache.rs

```diff
@@ -2,27 +2,373 @@
 
 use async_trait::async_trait;
 use clippy_utilities::OverflowArithmetic;
+use hashlink::LinkedHashSet;
 use lockfree_cuckoohash::{pin, LockFreeCuckooHash as HashMap};
-use std::collections::HashMap as StdHashMap;
-use std::sync::Arc;
-use tokio::sync::RwLock;
+use std::collections::{BTreeSet, HashMap as StdHashMap};
+use std::num::NonZeroUsize;
+use std::sync::{Arc, Weak};
+use std::time::Duration;
+use tokio::select;
+use tokio::sync::{mpsc, oneshot, RwLock};
+use tracing::warn;
 
 use super::policy::EvictPolicy;
 use super::{Block, BlockCoordinate, BlockId, Storage};
 use crate::async_fuse::fuse::protocol::INum;
 
-/// The file-level cache.
-type FileCache = Arc<RwLock<StdHashMap<BlockId, Block>>>;
-
 /// Merge the content from `src` to `dst`. This will set `dst` to be dirty.
 fn merge_two_blocks(src: &Block, dst: &mut Block) {
-    dst.set_dirty();
+    dst.set_dirty(true);
     dst.update(src);
 }
 
+/// A command sent from `MemoryCache` to the write back task.
+enum Command {
+    /// Store a block to the backend later.
+    Store {
+        /// The coordinate of the block
+        coord: BlockCoordinate,
+        /// The block
+        block: Block,
+        /// A sender to notify the `MemoryCache` that the block has been flushed to the backend
+        sender: oneshot::Sender<()>,
+    },
+    /// Flush the file specified immediately.
+    Flush(INum),
+    /// Flush all the files
+    FlushAll(oneshot::Sender<()>),
+}
+
+/// Functions for write back task.
+mod write_back_task {
+    use clippy_utilities::OverflowArithmetic;
+    use hashlink::LinkedHashSet;
+    use tokio::sync::oneshot;
+    use tracing::warn;
+
+    use std::{mem, sync::Arc, time::Duration};
+
+    use super::{Block, BlockId, EvictPolicy, INum, MemoryCache, SoftLimit, StdHashMap, Storage};
+    use crate::async_fuse::memfs::cache::BlockCoordinate;
+    /// Calculate a new period for write back task's interval.
+    pub(super) fn calculate_period(
+        pending_length: usize,
+        command_queue_limit: usize,
+        basic_period: Duration,
+    ) -> Duration {
+        if pending_length > command_queue_limit.overflow_div(4).overflow_mul(3) {
```

Comment:
Make these msgic numbers to constants.

Reply:
They are also fractions. I will write more comments to make it clearer, but I think there is no need to make them to constants.